src/utils/indexer.js

```diff
@@ -181,63 +201,68 @@ export const getXTZPriceInUSD = (usd, mutez) => {
   return parseFloat(convertMutezToXtz(mutez) * usd).toFixed(2);
 };
 
-export const getAllNFTHoldersInfo = () => {};
+// export const getAllNFTHoldersInfo = () => {};
 
 export const getNftInfoByXTZAddress = async (address = '') => {
-  const allTokens = await getAllNFTsMetadata();
-  const tokensOnOffer = await nftOnOffer();
+  try {
+    const allTokens = await getAllNFTsMetadata();
+    const tokensOnOffer = await nftOnOffer();
 
-  const response = await fetch(
-    `https://api.better-call.dev/v1/contract/${INDEXER_NETWORK}/${CONTRACT_ADDRESS}/storage`,
-  );
-  const result = await response.json();
-  const tokens = result[0].children.find(elm => elm.name === 'ledger');
+    const response = await fetch(
+      `https://api.better-call.dev/v1/contract/${INDEXER_NETWORK}/${CONTRACT_ADDRESS}/storage`,
+    );
+    const result = await response.json();
+    const tokens = result[0].children.find(elm => elm.name === 'ledger');
 
-  // const ledgerMetadata = await fetch(
-  //   `https://api.better-call.dev/v1/bigmap/${INDEXER_NETWORK}/${tokens.value}`,
-  // );
-  // const ledgerMetadataJSON = await ledgerMetadata.json();
+    let ledger = [];
+    let tk;
+    let offset = 0;
+    while (true) {
+      tk = await fetch(
+        `https://api.better-call.dev/v1/bigmap/${INDEXER_NETWORK}/${tokens.value}/keys?q=${address}&offset=${offset}`,
+      );
+      const res = await tk.json();
 
-  const tk = await fetch(
-    `https://api.better-call.dev/v1/bigmap/${INDEXER_NETWORK}/${tokens.value}/keys`,
-  );
+      if (res.length === 0) {
+        break;
+      }
 
-  const ledger = await tk.json();
+      ledger.push(...res);
+      offset += 10;
+    }
 
-  const owners = ledger.filter(elm => elm.data.value.value !== '0');
+    const owners = ledger.filter(elm => elm.data.value.value !== '0');
 
-  const allTokenHolders = owners.map(elm => {
-    return {
+    const tokenHolder = owners.map(elm => ({
       address: elm.data.key.children[0].value,
       tokenId: elm.data.key.children[1].value,
-    };
-  });
-
-  const tokenHolderUser = allTokenHolders.filter(el => el.address === address);
+    }));
+    console.log('holder', tokenHolder);
+    if (tokenHolder.length <= 0) {
+      return [];
+    }
+
+    const filtered = tokenHolder.map(elm => {
+      const nft = allTokens.find(element => element.tokenId == elm.tokenId);
+      const offer = tokensOnOffer.find(
+        element => element.tokenId == elm.tokenId,
+      );
 
-  if (tokenHolderUser.length <= 0) {
-    return [];
+      return {
+        address: elm.address,
+        tokenId: elm.tokenId,
+        uri: nft.uri,
+        isForSale: offer ? offer.isForSale : false,
+        saleValueInMutez: offer ? offer.saleValueInMutez : null,
+        seller: offer ? offer.seller : null,
+        offerDate: offer ? offer.timestamp : null,
+        owner: elm.address,
+      };
+    });
+    return filtered;
```

Comment:
There are two returns here that seem confusing. Please elaborate on what's happening here.

Reply:
The first return is the return inside `tokenHolder.map`.

The second return is what the function is returning.

Reply:
Got it.